### PR TITLE
Expert series articles

### PR DESCRIPTION
**Pull request by joenash**
This PR adds 3 articles to what we're currently calling the "expert series". These articles are:
- Choosing TCP or UDP: a guide for game developers
- Handling Session and Authentication within your games
- Modeling game mechanics for scalability

This PR is currently in a **Draft** state. Here is what is left to do:
- [x] Review draft images and create versions appropriate for live docs.
- [x] Add images to `images` folder and link within article.
- [x] Decide home for articles within nav and add to nav file.

### CODE REVIEW

File: docs/expert-tcp-udp.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+**Know what messages, if lost, would annoy the player** (or worse), what they’d never notice, and what’s in between. Each message’s place on the reliability gradient is key. If you know that prediction, interpolation, or other strategies can gracefully hide the occasional lost message, then that gives you more options for sending that message.
+
+## Stand on shoulders of giants
+
+If you’re thinking that this only begins to dig into thinking about net code, then you’re right. As you build out your games netcode, you’ll need to learn more about different aspects of prediction, interpolation, serialization, and beyond.
+
+Read Ruoyu Sun’s *[Game Networking Demystifie*d](https://ruoyusun.com/2019/03/28/game-networking-1.html), which takes a closer look at game states and client inputs and how they relate to cheating, server authority, and game genre.
+
+Another great read is *[Peeking into VALORANT's Netcod*e](https://technology.riotgames.com/news/peeking-valorants-netcode) from Riot Games, which goes in-depth on how product design goals can lead to specific decisions in netcode design.
```
Comment (lugehorsam) [inline docs/expert-tcp-udp.md:115]: asterisk typo here

File: docs/expert-tcp-udp.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+**Know what messages, if lost, would annoy the player** (or worse), what they’d never notice, and what’s in between. Each message’s place on the reliability gradient is key. If you know that prediction, interpolation, or other strategies can gracefully hide the occasional lost message, then that gives you more options for sending that message.
+
+## Stand on shoulders of giants
+
+If you’re thinking that this only begins to dig into thinking about net code, then you’re right. As you build out your games netcode, you’ll need to learn more about different aspects of prediction, interpolation, serialization, and beyond.
+
+Read Ruoyu Sun’s *[Game Networking Demystifie*d](https://ruoyusun.com/2019/03/28/game-networking-1.html), which takes a closer look at game states and client inputs and how they relate to cheating, server authority, and game genre.
```
Comment (lugehorsam) [inline docs/expert-tcp-udp.md:113]: asterisk typo here

File: docs/expert-tcp-udp.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+For points on the gradient between optional and required, there’s a strategy called *optionally reliable UDP*. This is when you build additional guarantees in your application, rather than relying on the protocol. In outline, this is sending UDP packets that require some response from the recipient; the message is re-sent until the expected response is received.
+
+But whether you choose UDP or TCP or optionally reliable UDP comes down to the *content* of your messages and what they mean to your game and players.
+
+## Start with your payloads 
+
+The content of your messages matters more than the vehicle they travel in. Start by focusing on your payloads, then figure out how to get them there.
+
+Before you start sending messages, make sure you**_ _****know what you need to send over the wire** (and what you don’t). For example, in an action game, your player’s movements will probably be sent to the server many times per second, while your player’s inventory status only needs to be sent when it changes. Smaller data structures are less likely to pay fragmentation penalties. The fastest packet is the one you never send.
```
Comment (lugehorsam) [inline docs/expert-tcp-udp.md:103]: I think an asterisk typo here unless I'm missing some special sequence.

Comment (joenash) [reply docs/expert-tcp-udp.md:103]: Almost definitely an error, it seems the Google Docs -> Markdown generator I used has some interesting asterisk quirks. Thanks for catching all of these @lugehorsam!

File: docs/expert-tcp-udp.md (hunk 1 above)
Comment (joenash) [reply docs/expert-tcp-udp.md:115]:
```suggestion
Another great read is *[Peeking into VALORANT's Netcode](https://technology.riotgames.com/news/peeking-valorants-netcode)* from Riot Games, which goes in-depth on how product design goals can lead to specific decisions in netcode design.
```

File: docs/expert-tcp-udp.md (hunk 2 above)
Comment (joenash) [reply docs/expert-tcp-udp.md:113]:
```suggestion
Read Ruoyu Sun’s [*Game Networking Demystified*](https://ruoyusun.com/2019/03/28/game-networking-1.html), which takes a closer look at game states and client inputs and how they relate to cheating, server authority, and game genre.
```

File: docs/expert-tcp-udp.md (hunk 3 above)
Comment (joenash) [reply docs/expert-tcp-udp.md:103]:
```suggestion
Before you start sending messages, make sure you **know what you need to send over the wire** (and what you don’t). For example, in an action game, your player’s movements will probably be sent to the server many times per second, while your player’s inventory status only needs to be sent when it changes. Smaller data structures are less likely to pay fragmentation penalties. The fastest packet is the one you never send.
```

File: docs/expert-auth.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+Other session types exist and have a role in developing authentication for your game (such as a network TCP session or a cryptographic TLS session), but they’re not essential to this article. We’ll assume these underlying protocols just work and focus more on player interactions.
+
+## Sessions overlap
+
+Players don’t generally think of being signed-in or about server connections. They think about playing and not playing. They don’t expect to enter their password or face down a "Connecting" screen every time they take a turn.
+
+As a consequence, session types overlap and depend on each other. This is how sign-ins and authentication support your game: they run longer or shorter than a play session, in ways that are (hopefully) transparent to the player. You can imagine your game being in many sessions at once: the player signed in, the player spending time playing the game, and the client in connection with Nakama.
+
+![image alt text](image_0.png)
```
Comment (joenash) [inline docs/expert-auth.md:45]:
```suggestion
![The overlap between player behaviours and session types](nakama-sessions.png)
```

File: docs/expert-modelling.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+Other events might trigger actions against the streak, too. For example, a player might have the choice to cancel a streak (perhaps to start on some other progression). This would have its own event handler.
+
+## Event hooks in Nakama
+
+In Nakama, we can implement this pattern with *[after hook*s](https://heroiclabs.com/docs/runtime-code-basics/#after-hook)*. *You can register a function that executes after each message received by the server. In that function, you can dispatch to more specific event handlers to level up players, award trophies, or, in this example, update a streak.
+
+In this case, you can write an after hook that updates the player’s streak progression whenever a client calls the `getAccount()` function that sends an out-of-band notification about the streak.
+
+![image alt text](image_0.png)In this case, we get an event trigger on an API that Nakama already provides and that you’re likely to use anyway: `getAccount()`. On the client side, there’s no additional code needed to trigger a streak update.
```
Comment (joenash) [inline docs/expert-modelling.md:128]:
```suggestion
![After hook updating a player's streak progression when getAccount function is called](nakama-after-hook.png)In this case, we get an event trigger on an API that Nakama already provides and that you’re likely to use anyway: `getAccount()`. On the client side, there’s no additional code needed to trigger a streak update.
```

File: docs/expert-tcp-udp.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+## Reliability is a gradient
+
+Transport layer protocols partly map to the gradient by setting distinct expectations for message delivery.
+
+UDP offers few reliability guarantees. It’s a more trusting protocol that expects that a message will be delivered. If it’s not delivered, the protocol won’t hear or do anything about it. It also doesn’t make any assumptions about when the message is delivered. UDP makes no effort to ensure the order of delivery. The first message you sent may not be the first received. When it comes to reliability, UDP is relaxed.
+
+TCP offers robust reliability guarantees. It goes to some effort to confirm that a connection between the sender and recipient is established, that the recipient is prepared to receive messages, and, after the messages are sent, that they are delivered completely and in order. When it comes to reliability, TCP is a strict micromanager.
+
+![image alt text](image_1.png)
```
Comment (joenash) [inline docs/expert-tcp-udp.md:49]:
```suggestion
![UDP vs TCP for reliability and overhead](nakama-tcp-vs-udp.png)
```

File: docs/expert-tcp-udp.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+### A Tale of Four Messages
+
+Under good network conditions, it’s actually quite hard to tell UDP and TCP apart in terms of reliability. But if conditions are poor, it’s easier to see how reliability tradeoffs emerge. To illustrate, suppose your game needs to send four messages sequentially (but let’s ignore their contents and importance, for now).
+
+In good conditions, messages arrive intact, promptly, in-order, and just once. In poorer conditions, none of these assumptions hold. How do the different protocols cope?
+
+With UDP, send the four messages, M1, M2, M3, and M4, each one after the other. What happens to each?
+
+![image alt text](image_2.png)
```
Comment (joenash) [inline docs/expert-tcp-udp.md:59]:
```suggestion
![Visualization of operation described below.](nakama-udp.png)
```

File: docs/expert-tcp-udp.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+* M3 arrives first!
+
+* M4 arrives, but a router along the way spontaneously duplicated the message in transit, so it arrives *twice*, once in second place and once in third.
+
+As far as UDP is concerned, this is how it’s supposed to happen. It’s "best effort" but at least it’s low overhead. The protocol doesn’t require the receiver to notify the sender of any problems. If anything is to be done about this network hiccup—to make sense of out-of-order delivery, repeated delivery, or the lost message the receiver didn’t even know about—it has to be done at the application layer, in your game.
+
+Let’s rewind things and play this scenario out again over TCP. TCP is a more complex protocol and we’ll have to tell a more complex story. But let’s send those messages again, assuming that we’ve already completed the three-way handshake and we’re in the ESTABLISHED state. What happens to each?
+
+![image alt text](image_3.png)
```
Comment (joenash) [inline docs/expert-tcp-udp.md:73]:
```suggestion
![Visualization of operations described below.](nakama-tcp.png)
```

**Pull request closed without merging**